# ML-07 — Baseline Action Score and Top-20 Review

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/prarthanamahesh21-hub/flyrank-ml-internship/blob/main/work/notebooks/w04_baseline_score.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [1]:
import pandas as pd
import numpy as np

print("Week 4 Baseline Action Score")
print("Setup complete.")

Week 4 Baseline Action Score
Setup complete.


In [2]:
import os
from google.colab import userdata
from datasets import load_dataset

HF_TOKEN = userdata.get("HF_TOKEN")

print("HF token loaded:", HF_TOKEN is not None)

HF token loaded: True


In [3]:
from datasets import load_dataset

dim_content = load_dataset(
    "FlyRank/internship-warehouse",
    "dim_content",
    token=HF_TOKEN
)

fact_performance = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance",
    token=HF_TOKEN
)

fact_query = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_query_90d",
    token=HF_TOKEN
)

print(dim_content)
print(fact_performance)
print(fact_query)

README.md:   0%|          | 0.00/3.04k [00:00<?, ?B/s]

dim_content.parquet: reconstructing file:   0%|          |  0.00B / 19.6MB            

dim_content.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/519606 [00:00<?, ? examples/s]

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 19.6kB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 4.41MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.29MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.22MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 8.93MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 72.0MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  624kB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 86.2MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 89.0MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 1.45MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 21.6MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  134MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 7.12MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 2.62MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 90.3MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  149MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  146MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/78835655 [00:00<?, ? examples/s]

Loading dataset shards:   0%|          | 0/39 [00:00<?, ?it/s]

fact_content_query_90d.parquet: reconstructing file:   0%|          |  0.00B / 60.7MB            

fact_content_query_90d.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/2414248 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['client_hash_id', 'content_hash_id', 'keyword_hash_id', 'url_hash_id', 'keyword_char_count', 'keyword_token_count', 'url_char_count', 'content_created_date', 'content_updated_date', 'content_type', 'search_volume', 'competition', 'competition_level', 'cpc', 'main_intent', 'backlinks', 'category_count', 'keyword_created_date', 'provider_used', 'model_used', 'char_count', 'word_count', 'last_optimized_date', 'optimization_eligible_date', 'is_published', 'is_deleted'],
        num_rows: 519606
    })
})
DatasetDict({
    train: Dataset({
        features: ['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'ses

In [ ]:
features_march[
    [
        "content_age_days",
        "gsc_clicks_90d"
    ]
].describe()

In [5]:
import pandas as pd
import duckdb

content = dim_content["train"]
performance = fact_performance["train"]
query_90d = fact_query["train"]

print("Content rows:", len(content))
print("Performance rows:", len(performance))
print("Query 90d rows:", len(query_90d))

Content rows: 519606
Performance rows: 78835655
Query 90d rows: 2414248


In [6]:
con = duckdb.connect()

con.register("dim_content", content._data.table)
con.register(
    "fact_content_daily_performance",
    performance._data.table
)
con.register(
    "fact_content_query_90d",
    query_90d._data.table
)

print("DuckDB setup complete")

DuckDB setup complete


In [9]:
print(query_90d.column_names)

['client_hash_id', 'content_hash_id', 'query_hash_id', 'query_char_count', 'query_token_count', 'window_start', 'window_end', 'impressions_90d', 'clicks_90d', 'impressions_last30', 'clicks_last30', 'impressions_prev30', 'clicks_prev30', 'avg_position_90d', 'avg_position_last30', 'avg_position_prev30', 'content_total_impressions_90d', 'content_visible_query_count', 'rare_query_count', 'rare_impressions_share', 'anonymized_impressions_share']


In [10]:
print(content.column_names)

['client_hash_id', 'content_hash_id', 'keyword_hash_id', 'url_hash_id', 'keyword_char_count', 'keyword_token_count', 'url_char_count', 'content_created_date', 'content_updated_date', 'content_type', 'search_volume', 'competition', 'competition_level', 'cpc', 'main_intent', 'backlinks', 'category_count', 'keyword_created_date', 'provider_used', 'model_used', 'char_count', 'word_count', 'last_optimized_date', 'optimization_eligible_date', 'is_published', 'is_deleted']


In [11]:
# Build a leakage-safe Week 4 feature frame
# using the existing 90-day query table.

features_w04 = con.sql("""
WITH query_by_content AS (
    SELECT
        content_hash_id,
        SUM(COALESCE(impressions_90d, 0)) AS gsc_impressions_90d,
        SUM(COALESCE(clicks_90d, 0)) AS gsc_clicks_90d,
        SUM(COALESCE(content_visible_query_count, 0)) AS visible_query_count
    FROM fact_content_query_90d
    GROUP BY content_hash_id
)

SELECT
    c.content_hash_id,

    DATE_DIFF(
        'day',
        CAST(c.content_created_date AS DATE),
        DATE '2026-03-01'
    ) AS content_age_days,

    c.word_count,

    COALESCE(q.gsc_impressions_90d, 0) AS gsc_impressions_90d,
    COALESCE(q.gsc_clicks_90d, 0) AS gsc_clicks_90d,
    COALESCE(q.visible_query_count, 0) AS visible_query_count

FROM dim_content c

LEFT JOIN query_by_content q
    ON c.content_hash_id = q.content_hash_id

WHERE c.content_created_date < DATE '2026-03-01'
  AND c.is_published IS TRUE
  AND c.is_deleted IS FALSE
""").df()

print("Feature frame shape:", features_w04.shape)
features_w04.head()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Feature frame shape: (303321, 6)


,content_hash_id,content_age_days,word_count,gsc_impressions_90d,gsc_clicks_90d,visible_query_count
0,content_00014efc121d911d,214,<NA>,70.0,1.0,1.0
1,content_000f4b73532b9e9d,213,<NA>,21.0,0.0,1.0
2,content_001689f50648eef1,163,2464,827.0,1.0,324.0
3,content_0026971d958baec2,163,<NA>,91.0,0.0,16.0
4,content_002bf62039e49f93,214,<NA>,88.0,0.0,16.0


In [12]:
print(features_w04.columns.tolist())

print("\nMissing values:")
print(features_w04[
    [
        "content_age_days",
        "gsc_clicks_90d",
        "gsc_impressions_90d"
    ]
].isna().sum())

print("\nSummary:")
print(features_w04[
    [
        "content_age_days",
        "gsc_clicks_90d",
        "gsc_impressions_90d"
    ]
].describe())

['content_hash_id', 'content_age_days', 'word_count', 'gsc_impressions_90d', 'gsc_clicks_90d', 'visible_query_count']

Missing values:
content_age_days       0
gsc_clicks_90d         0
gsc_impressions_90d    0
dtype: int64

Summary:
       content_age_days  gsc_clicks_90d  gsc_impressions_90d
count     303321.000000   303321.000000        303321.000000
mean         197.188226        1.164196           557.982022
std          109.097729       15.544630          4641.318895
min            1.000000        0.000000             0.000000
25%          128.000000        0.000000             0.000000
50%          204.000000        0.000000             0.000000
75%          271.000000        0.000000            40.000000
max          464.000000     2422.000000        436607.000000


In [ ]:
print(features_march.columns.tolist())

In [13]:
age_check = features_w04.copy()

age_check["age_bucket"] = pd.cut(
    age_check["content_age_days"],
    bins=[-1, 90, 180, 365, np.inf],
    labels=["0-90 days", "91-180 days", "181-365 days", "365+ days"]
)

age_table = (
    age_check
    .groupby("age_bucket", observed=False)
    .agg(
        n=("content_hash_id", "size"),
        median_impressions=("gsc_impressions_90d", "median"),
        median_clicks=("gsc_clicks_90d", "median")
    )
    .reset_index()
)

age_table

,age_bucket,n,median_impressions,median_clicks
0,0-90 days,66694,0.0,0.0
1,91-180 days,56173,0.0,0.0
2,181-365 days,160368,0.0,0.0
3,365+ days,20086,0.0,0.0


In [14]:
volume_check = features_w04.copy()

volume_check["impression_bucket"] = pd.cut(
    volume_check["gsc_impressions_90d"],
    bins=[-1, 0, 10, 100, 1000, np.inf],
    labels=["0", "1-10", "11-100", "101-1000", "1000+"]
)

volume_table = (
    volume_check
    .groupby("impression_bucket", observed=False)
    .agg(
        n=("content_hash_id", "size"),
        median_age_days=("content_age_days", "median"),
        median_clicks=("gsc_clicks_90d", "median")
    )
    .reset_index()
)

volume_table

,impression_bucket,n,median_age_days,median_clicks
0,0,207610,209.0,0.0
1,1-10,2064,163.0,0.0
2,11-100,33162,164.0,0.0
3,101-1000,39183,186.0,0.0
4,1000+,21302,165.0,4.0


## 1. My rule and its reason codes

*Write the rule in plain words first. Then the reason codes it can output.*


**Rule:** Prioritize pages for a possible content refresh when they are older and have meaningful search visibility. Content age is used as a staleness signal, while 90-day Google Search Console impressions are used as a search-opportunity signal.

**Signal check verdicts:**

* **`content_age_days` — MIXED:** The age bucket check did not show higher median search activity for older pages. Median impressions and clicks were zero across all age buckets, so age alone does not demonstrate higher search activity.
* **`gsc_impressions_90d` — CONFIRMED:** The 1,000+ impression bucket had a median of 4 clicks, while the lower-impression buckets had a median of 0 clicks. This supports using impressions as a directional measure of search visibility.

**Reason code:**

* `STALE_HIGH_VALUE` — The page is relatively old and has meaningful search visibility, so it receives higher priority for refresh review.


In [3]:
from google.colab import userdata

HF_TOKEN = userdata.get("HF_TOKEN")

print("HF token loaded:", HF_TOKEN is not None)

HF token loaded: True


In [4]:
from datasets import load_dataset

dim_content = load_dataset(
    "FlyRank/internship-warehouse",
    "dim_content",
    token=HF_TOKEN
)

fact_query = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_query_90d",
    token=HF_TOKEN
)

print("Datasets loaded.")

README.md:   0%|          | 0.00/3.04k [00:00<?, ?B/s]

dim_content.parquet: reconstructing file:   0%|          |  0.00B / 19.6MB            

dim_content.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/519606 [00:00<?, ? examples/s]

fact_content_query_90d.parquet: reconstructing file:   0%|          |  0.00B / 60.7MB            

fact_content_query_90d.parquet: downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/2414248 [00:00<?, ? examples/s]

Datasets loaded.


In [5]:
import duckdb

content = dim_content["train"]
query_90d = fact_query["train"]

con = duckdb.connect()

con.register("dim_content", content._data.table)
con.register("fact_content_query_90d", query_90d._data.table)

print("DuckDB setup complete")
print("Content rows:", len(content))
print("Query 90d rows:", len(query_90d))

DuckDB setup complete
Content rows: 519606
Query 90d rows: 2414248


In [6]:
features_w04 = con.sql("""
WITH query_by_content AS (
    SELECT
        content_hash_id,
        SUM(COALESCE(impressions_90d, 0)) AS gsc_impressions_90d,
        SUM(COALESCE(clicks_90d, 0)) AS gsc_clicks_90d,
        SUM(COALESCE(content_visible_query_count, 0)) AS visible_query_count
    FROM fact_content_query_90d
    GROUP BY content_hash_id
)

SELECT
    c.content_hash_id,

    DATE_DIFF(
        'day',
        CAST(c.content_created_date AS DATE),
        DATE '2026-03-01'
    ) AS content_age_days,

    c.word_count,

    COALESCE(q.gsc_impressions_90d, 0) AS gsc_impressions_90d,
    COALESCE(q.gsc_clicks_90d, 0) AS gsc_clicks_90d,
    COALESCE(q.visible_query_count, 0) AS visible_query_count

FROM dim_content c

LEFT JOIN query_by_content q
    ON c.content_hash_id = q.content_hash_id

WHERE c.content_created_date < DATE '2026-03-01'
  AND c.is_published IS TRUE
  AND c.is_deleted IS FALSE
""").df()

print("Feature frame shape:", features_w04.shape)

Feature frame shape: (303321, 6)


In [11]:
import pandas as pd
import numpy as np

print("Pandas and NumPy ready.")

Pandas and NumPy ready.


In [12]:
baseline = features_w04.copy()

## 2. Build the ranked queue (writes the CSV)

*Code the score, rank everything, write work/outputs/baseline_action_score.csv.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [8]:
import pandas as pd
import numpy as np

print("Pandas and NumPy ready.")

Pandas and NumPy ready.


In [10]:
baseline = features_w04.copy()

In [13]:
print("Baseline rows:", len(baseline))
print("Baseline columns:", baseline.columns.tolist())

Baseline rows: 303321
Baseline columns: ['content_hash_id', 'content_age_days', 'word_count', 'gsc_impressions_90d', 'gsc_clicks_90d', 'visible_query_count']


In [18]:
baseline["age_points"] = pd.cut(
    baseline["content_age_days"],
    bins=[-1, 90, 180, 365, np.inf],
    labels=[0, 1, 2, 3]
).astype(int)

print("Age points created.")
print(baseline["age_points"].value_counts().sort_index())

Age points created.
age_points
0     66694
1     56173
2    160368
3     20086
Name: count, dtype: int64


In [20]:
baseline["score"] = (
    baseline["age_points"] +
    baseline["impression_points"]
)

print("Score created.")
print(baseline["score"].describe())

Score created.
count    303321.000000
mean          2.335189
std           1.552906
min           0.000000
25%           2.000000
50%           2.000000
75%           3.000000
max           7.000000
Name: score, dtype: float64


In [19]:
baseline["impression_points"] = pd.cut(
    baseline["gsc_impressions_90d"],
    bins=[-1, 0, 10, 100, 1000, np.inf],
    labels=[0, 1, 2, 3, 4]
).astype(int)

print("Impression points created.")
print(baseline["impression_points"].value_counts().sort_index())

Impression points created.
impression_points
0    207610
1      2064
2     33162
3     39183
4     21302
Name: count, dtype: int64


In [21]:
baseline["score"] = (
    baseline["age_points"] +
    baseline["impression_points"]
)

print("Score created.")
print(baseline["score"].describe())

Score created.
count    303321.000000
mean          2.335189
std           1.552906
min           0.000000
25%           2.000000
50%           2.000000
75%           3.000000
max           7.000000
Name: score, dtype: float64


In [22]:
baseline["reason_code"] = "STALE_HIGH_VALUE"

baseline["action"] = np.where(
    baseline["score"] >= 5,
    "Refresh",
    "Monitor"
)

print("Reason code and action created.")
print(baseline["action"].value_counts())

Reason code and action created.
action
Monitor    266202
Refresh     37119
Name: count, dtype: int64


In [23]:
baseline = baseline.sort_values(
    ["score", "content_age_days", "gsc_impressions_90d"],
    ascending=[False, False, False]
).reset_index(drop=True)

baseline["rank"] = np.arange(1, len(baseline) + 1)

print("Ranking complete.")
print("Total ranked pages:", len(baseline))

Ranking complete.
Total ranked pages: 303321


In [24]:
baseline_output = baseline[
    [
        "rank",
        "content_hash_id",
        "score",
        "reason_code",
        "action",
        "content_age_days",
        "gsc_impressions_90d",
        "gsc_clicks_90d"
    ]
]

baseline_output.head(10)

,rank,content_hash_id,score,reason_code,action,content_age_days,gsc_impressions_90d,gsc_clicks_90d
0,1,content_787dc8aeae12cf0c,7,STALE_HIGH_VALUE,Refresh,445,139056.0,491.0
1,2,content_93ece6825477a6fe,7,STALE_HIGH_VALUE,Refresh,445,94572.0,194.0
2,3,content_896b10e4029fa236,7,STALE_HIGH_VALUE,Refresh,445,58358.0,5.0
3,4,content_ffa4764d05b5f507,7,STALE_HIGH_VALUE,Refresh,445,58167.0,0.0
4,5,content_75c1e7c3c6d66ad7,7,STALE_HIGH_VALUE,Refresh,445,52845.0,24.0
5,6,content_823409eedd0827e9,7,STALE_HIGH_VALUE,Refresh,445,51281.0,0.0
6,7,content_3d6b3b02182ac879,7,STALE_HIGH_VALUE,Refresh,445,44736.0,114.0
7,8,content_375caf7085fbd383,7,STALE_HIGH_VALUE,Refresh,445,44256.0,0.0
8,9,content_7fab1fdf9219e29b,7,STALE_HIGH_VALUE,Refresh,445,32571.0,61.0
9,10,content_e01fcca334199f7a,7,STALE_HIGH_VALUE,Refresh,445,31118.0,197.0


In [15]:
# Build the baseline score

baseline = features_w04.copy()

# Age points
baseline["age_points"] = pd.cut(
    baseline["content_age_days"],
    bins=[-1, 90, 180, 365, np.inf],
    labels=[0, 1, 2, 3]
).astype(int)

# Search visibility points
baseline["impression_points"] = pd.cut(
    baseline["gsc_impressions_90d"],
    bins=[-1, 0, 10, 100, 1000, np.inf],
    labels=[0, 1, 2, 3, 4]
).astype(int)

# Final score
baseline["score"] = (
    baseline["age_points"] +
    baseline["impression_points"]
)

# One reason code
baseline["reason_code"] = "STALE_HIGH_VALUE"

# Action label
baseline["action"] = np.where(
    baseline["score"] >= 5,
    "Refresh",
    "Monitor"
)

# Rank pages
baseline = baseline.sort_values(
    ["score", "content_age_days", "gsc_impressions_90d"],
    ascending=[False, False, False]
).reset_index(drop=True)

baseline["rank"] = np.arange(1, len(baseline) + 1)

# Final output columns
baseline_output = baseline[
    [
        "rank",
        "content_hash_id",
        "score",
        "reason_code",
        "action",
        "content_age_days",
        "gsc_impressions_90d",
        "gsc_clicks_90d"
    ]
]

print("Ranked pages:", len(baseline_output))
print("\nTop 10:")
print(baseline_output.head(10).to_string(index=False))

Ranked pages: 303321

Top 10:
 rank          content_hash_id  score      reason_code  action  content_age_days  gsc_impressions_90d  gsc_clicks_90d
    1 content_787dc8aeae12cf0c      7 STALE_HIGH_VALUE Refresh               445             139056.0           491.0
    2 content_93ece6825477a6fe      7 STALE_HIGH_VALUE Refresh               445              94572.0           194.0
    3 content_896b10e4029fa236      7 STALE_HIGH_VALUE Refresh               445              58358.0             5.0
    4 content_ffa4764d05b5f507      7 STALE_HIGH_VALUE Refresh               445              58167.0             0.0
    5 content_75c1e7c3c6d66ad7      7 STALE_HIGH_VALUE Refresh               445              52845.0            24.0
    6 content_823409eedd0827e9      7 STALE_HIGH_VALUE Refresh               445              51281.0             0.0
    7 content_3d6b3b02182ac879      7 STALE_HIGH_VALUE Refresh               445              44736.0           114.0
    8 content_375caf7085fb

In [25]:
import os

os.makedirs("work/outputs", exist_ok=True)

output_path = "work/outputs/baseline_action_score.csv"

baseline_output.to_csv(
    output_path,
    index=False
)

print("CSV written to:", output_path)
print("Rows written:", len(baseline_output))

CSV written to: work/outputs/baseline_action_score.csv
Rows written: 303321


In [26]:
import os

os.makedirs("work/outputs", exist_ok=True)

output_path = "work/outputs/baseline_action_score.csv"

baseline_output.to_csv(
    output_path,
    index=False
)

print("CSV written to:", output_path)
print("Rows written:", len(baseline_output))

CSV written to: work/outputs/baseline_action_score.csv
Rows written: 303321


## 3. Top-20 review

*For each of the top 20: action, reason code, confidence note, and what would make it wrong.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [27]:
top20 = baseline_output.head(20).copy()

top20

,rank,content_hash_id,score,reason_code,action,content_age_days,gsc_impressions_90d,gsc_clicks_90d
0,1,content_787dc8aeae12cf0c,7,STALE_HIGH_VALUE,Refresh,445,139056.0,491.0
1,2,content_93ece6825477a6fe,7,STALE_HIGH_VALUE,Refresh,445,94572.0,194.0
2,3,content_896b10e4029fa236,7,STALE_HIGH_VALUE,Refresh,445,58358.0,5.0
3,4,content_ffa4764d05b5f507,7,STALE_HIGH_VALUE,Refresh,445,58167.0,0.0
4,5,content_75c1e7c3c6d66ad7,7,STALE_HIGH_VALUE,Refresh,445,52845.0,24.0
5,6,content_823409eedd0827e9,7,STALE_HIGH_VALUE,Refresh,445,51281.0,0.0
6,7,content_3d6b3b02182ac879,7,STALE_HIGH_VALUE,Refresh,445,44736.0,114.0
7,8,content_375caf7085fbd383,7,STALE_HIGH_VALUE,Refresh,445,44256.0,0.0
8,9,content_7fab1fdf9219e29b,7,STALE_HIGH_VALUE,Refresh,445,32571.0,61.0
9,10,content_e01fcca334199f7a,7,STALE_HIGH_VALUE,Refresh,445,31118.0,197.0


## 3. Top-20 review

The baseline ranks pages using content age and 90-day search impressions. All top-20 pages receive the maximum score of 7 because they are 445 days old and have more than 1,000 impressions. They therefore receive the `Refresh` action and `STALE_HIGH_VALUE` reason code. Confidence is higher when the page also has measurable clicks, while pages with zero clicks should be treated more cautiously.

| Rank | Action  | Reason code      | Confidence note                                                                       | What would make it wrong                                                                                                               |
| ---: | ------- | ---------------- | ------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
|    1 | Refresh | STALE_HIGH_VALUE | High visibility (139,056 impressions) and 491 clicks; strong review priority.         | The page may have been recently refreshed despite its creation age, or impressions may not represent a meaningful refresh opportunity. |
|    2 | Refresh | STALE_HIGH_VALUE | High visibility (94,572 impressions) and 194 clicks.                                  | The page may already be current or its search demand may have changed.                                                                 |
|    3 | Refresh | STALE_HIGH_VALUE | High visibility (58,358 impressions) but only 5 clicks; opportunity is less certain.  | The impressions may come from poor-position or low-intent searches, making refresh less useful.                                        |
|    4 | Refresh | STALE_HIGH_VALUE | High visibility (58,167 impressions) but zero clicks; needs cautious review.          | The page may have impressions without meaningful search opportunity, or another issue may explain the lack of clicks.                  |
|    5 | Refresh | STALE_HIGH_VALUE | High visibility (52,845 impressions) and 24 clicks.                                   | The page may already satisfy users despite its age, or the impressions may be unrelated to refresh potential.                          |
|    6 | Refresh | STALE_HIGH_VALUE | High visibility (51,281 impressions) but zero clicks; lower confidence.               | Zero clicks could indicate weak relevance or SERP positioning rather than content staleness.                                           |
|    7 | Refresh | STALE_HIGH_VALUE | High visibility (44,736 impressions) and 114 clicks.                                  | The page may be performing adequately and not need a content refresh.                                                                  |
|    8 | Refresh | STALE_HIGH_VALUE | High visibility (44,256 impressions) but zero clicks; cautious priority.              | High impressions may not translate into useful traffic or refresh opportunity.                                                         |
|    9 | Refresh | STALE_HIGH_VALUE | High visibility (32,571 impressions) and 61 clicks.                                   | The page may already be sufficiently useful, or its performance may have another explanation.                                          |
|   10 | Refresh | STALE_HIGH_VALUE | High visibility (31,118 impressions) and 197 clicks; strong search engagement signal. | The page may have changed recently or the opportunity may require a different action than refresh.                                     |
|   11 | Refresh | STALE_HIGH_VALUE | High visibility (29,293 impressions) and 45 clicks.                                   | The page may not have a content-quality problem despite being old.                                                                     |
|   12 | Refresh | STALE_HIGH_VALUE | High visibility (24,915 impressions) and 28 clicks.                                   | Search visibility may not indicate that content changes would improve performance.                                                     |
|   13 | Refresh | STALE_HIGH_VALUE | High visibility (21,548 impressions) but only 2 clicks; lower confidence.             | The impressions may reflect weak rankings or mismatched search intent rather than stale content.                                       |
|   14 | Refresh | STALE_HIGH_VALUE | High visibility (21,203 impressions) and 101 clicks.                                  | The page may already be performing well enough that refresh effort would have limited benefit.                                         |
|   15 | Refresh | STALE_HIGH_VALUE | High visibility (19,812 impressions) and 18 clicks.                                   | The page may be old but still accurate and useful, so age alone may not justify refresh.                                               |
|   16 | Refresh | STALE_HIGH_VALUE | High visibility (17,415 impressions) and 35 clicks.                                   | The observed visibility may not be caused by outdated content.                                                                         |
|   17 | Refresh | STALE_HIGH_VALUE | High visibility (16,646 impressions) and 25 clicks.                                   | A refresh may not improve performance if the underlying search demand or intent has changed.                                           |
|   18 | Refresh | STALE_HIGH_VALUE | High visibility (16,000 impressions) and 78 clicks.                                   | The page may already be effective and may need monitoring rather than a refresh.                                                       |
|   19 | Refresh | STALE_HIGH_VALUE | High visibility (15,284 impressions) and 10 clicks; moderate-to-low confidence.       | The impressions may not represent a strong opportunity for content improvement.                                                        |
|   20 | Refresh | STALE_HIGH_VALUE | High visibility (14,611 impressions) and 81 clicks.                                   | The page may already be meeting its search need, making refresh unnecessary.                                                           |


## 4. Weak picks + leakage check

*Which picks look wrong and why? Confirm no product flags or future windows leaked in.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Weak picks + leakage check

**Weak picks:** The weakest picks in the top 20 are ranks 4, 6, 8, and 13 because they have high impressions but zero or very few clicks. The baseline treats high impressions as search visibility, but these pages may not represent strong refresh opportunities. This shows that the rule should be treated as a prioritization baseline rather than a claim that every selected page needs refreshing.

**Leakage check:** The baseline score uses only `content_age_days` and `gsc_impressions_90d`, both measured before the March 2026 decision point. No April 2026 outcome label, future-window clicks, or product flags are used in the baseline score or action assignment. The April label used elsewhere in the notebook is not an input to this baseline rule.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.